modules/images/image-loading-optimization/class-ilo-grouped-url-metrics.php

```diff
@@ -0,0 +1,295 @@
+	 *                                      metrics with viewports on either side of the breakpoint boundaries.
+	 */
+	private function group_url_metrics_by_breakpoint( array $url_metrics ): array {
+
+		// Convert breakpoint max widths into viewport minimum widths.
+		$minimum_viewport_widths = array_map(
+			static function ( $breakpoint ) {
+				return $breakpoint + 1;
+			},
+			$this->breakpoints
+		);
+
+		$grouped = array_fill_keys( array_merge( array( 0 ), $minimum_viewport_widths ), array() );
+
+		foreach ( $url_metrics as $url_metric ) {
+			$current_minimum_viewport = 0;
+			foreach ( $minimum_viewport_widths as $viewport_minimum_width ) {
+				if ( $url_metric->get_viewport()['width'] > $viewport_minimum_width ) {
+					$current_minimum_viewport = $viewport_minimum_width;
+				} else {
+					break;
+				}
+			}
+
+			$grouped[ $current_minimum_viewport ][] = $url_metric;
+		}
+		return $grouped;
+	}
+
+	/**
+	 * Gets needed minimum viewport widths.
+	 *
+	 * @since n.e.x.t
+	 * @access private
+	 *
+	 * @param float $current_time Current time, defaults to `microtime(true)`.
+	 * @return array<int, array{int, bool}> Array of tuples mapping minimum viewport width to whether URL metric(s) are needed.
+	 */
+	public function get_needed_minimum_viewport_widths( float $current_time = null ): array {
+		if ( null === $current_time ) {
+			$current_time = microtime( true );
+		}
+
+		$needed_minimum_viewport_widths = array();
+		foreach ( $this->groups as $minimum_viewport_width => $viewport_url_metrics ) {
+			$needs_url_metrics = false;
+			if ( count( $viewport_url_metrics ) < $this->sample_size ) {
+				$needs_url_metrics = true;
+			} else {
+				foreach ( $viewport_url_metrics as $url_metric ) {
+					if ( $url_metric->get_timestamp() + $this->freshness_ttl < $current_time ) {
+						$needs_url_metrics = true;
+						break;
+					}
+				}
+			}
+			$needed_minimum_viewport_widths[] = array(
+				$minimum_viewport_width,
+				$needs_url_metrics,
+			);
+		}
+
+		return $needed_minimum_viewport_widths;
+	}
+
+	/**
+	 * Gets the LCP element for each breakpoint.
+	 *
+	 * The array keys are the minimum viewport width required for the element to be LCP. If there are URL metrics for a
+	 * given breakpoint and yet there is no supported LCP element, then the array value is `false`. (Currently only IMG is
+	 * a supported LCP element.) If there is a supported LCP element at the breakpoint, then the array value is an array
+	 * representing that element, including its breadcrumbs. If two adjoining breakpoints have the same value, then the
+	 * latter is dropped.
+	 *
+	 * @since n.e.x.t
+	 * @access private
+	 *
+	 * @return array LCP elements keyed by its minimum viewport width. If there is no supported LCP element at a breakpoint, then `false` is used.
+	 */
+	public function get_lcp_elements_by_minimum_viewport_widths(): array {
```

Comment:
I would prefer to leave this method out of the class and keep as a standalone function for now. It is more specific than all other methods in the class in that it is tied to LCP related logic, whereas the rest of this class is more generic and reusable.

Reply:
OK. Moved back in bdeaaccd.